core/src/main/resources/lib/form/class-entry.jelly

```diff
@@ -36,12 +46,14 @@ THE SOFTWARE.
   </st:documentation>
   <j:set var="clazz" value="${attrs.clazz ?: attrs.descriptor.clazz.name}" />
   <f:invisibleEntry>
-    <!-- Legacy: Remove once plugins have been staged onto $class -->
-    <input type="hidden" name="stapler-class" value="${clazz}" />
-    <!-- Legacy: Remove once plugins have been staged onto $class -->
-    <j:if test="${attrs.descriptor != null}">
-      <input type="hidden" name="kind" value="${attrs.descriptor.id}" />
-    </j:if>
-    <input type="hidden" name="$class" value="${clazz}" />
+    <j:choose>
+      <j:when test="${descriptor != null and descriptor.id != clazz}">
+        <input type="hidden" name="kind" value="${attrs.descriptor.id}" />
+      </j:when>
```

Comment:
[JENKINS-28114 NodeJsInstaller exception](https://issues.jenkins-ci.org/browse/JENKINS-28114) is caused for here.
* `NodeJsInstaller` overrides `Descriptor#getId` (https://github.com/jenkinsci/nodejs-plugin/blob/nodejs-0.2.1/src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstaller.java#L383). This results no `$class` nor `stapler-class` as `descriptor.id != clazz`
* An exception occurs when the data binding is handled by stapler, as stapler never recognizes `kind`.